src/OrchardCore.Modules/OrchardCore.AutoSetup/AutoSetupMiddleware.cs

```diff
@@ -0,0 +1,182 @@
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using OrchardCore.Environment.Shell;
+using OrchardCore.Environment.Shell.Scope;
+using OrchardCore.Setup.Services;
+
+namespace OrchardCore.AutoSetup
+{
+    using OrchardCore.Abstractions.Setup;
+
+    /// <summary>
+    /// The auto setup middleware.
+    /// </summary>
+    public class AutoSetupMiddleware
+    {
+        /// <summary>
+        /// The next middleware in the execution pipeline.
+        /// </summary>
+        private readonly RequestDelegate _next;
+
+        /// <summary>
+        /// The Shell settings manager.
+        /// </summary>
+        private readonly IShellSettingsManager _shellSettingsManager;
+
+        /// <summary>
+        /// The auto-setup options.
+        /// </summary>
+        private readonly IOptions<AutoSetupOptions> _setupOptions;
+
+        /// <summary>
+        /// The _logger.
+        /// </summary>
+        private readonly ILogger<AutoSetupMiddleware> _logger;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AutoSetupMiddleware"/> class.
+        /// </summary>
+        /// <param name="next"> The next middleware in the execution pipeline. </param>
+        /// <param name="shellSettingsManager">The Shell settings manager</param>
+        /// <param name="setupOptions"> The auto-setup Options. </param>
+        /// <param name="logger"> The logger. </param>
+        public AutoSetupMiddleware(
+            RequestDelegate next,
+            IShellSettingsManager shellSettingsManager,
+            IOptions<AutoSetupOptions> setupOptions,
+            ILogger<AutoSetupMiddleware> logger)
+        {
+            _logger = logger;
+            _next = next;
+            _shellSettingsManager = shellSettingsManager;
+            _setupOptions = setupOptions;
+        }
+
+        /// <summary>
+        /// The middleware auto-setup invoke.
+        /// </summary>
+        /// <param name="httpContext">
+        /// The http context.
+        /// </param>
+        /// <returns>
+        /// The <see cref="Task"/>.
+        /// </returns>
+        public async Task InvokeAsync(HttpContext httpContext)
+        {
+            var currentShellSettings = ShellScope.Context.Settings;
+            var setupService = httpContext.RequestServices.GetRequiredService<ISetupService>();
+
+            _logger.LogInformation("AutoSetup is initializing the site");
+            if (await SetupTenant(setupService, _setupOptions.Value.RootTenant, currentShellSettings))
+            {
+                foreach (var tenant in _setupOptions.Value.SubTenants)
+                {
+                    if(!await SetupTenant(setupService, tenant, CreateTenantSettings(tenant)))
+                    {
+                        break;
+                    }
+                }
+
+                httpContext.Response.Redirect("/");
```

Comment:
~/ so it also works with the prefix of the default tenant

Reply:
Fixed

Reply:
Reverting it back, when I use redirection to ~/ it redirects me to https://localhost:5001/~/ 404

Reply:
@rserj 

Yes, `~/` works when using Redirect (returning a RedirectResult) from a controller, because when executing the result the url helper.Content() is called and it uses the pathBase.

The pathBase contains an eventual virtual folder, and at this point it already contains the tenant prefix, the default tenant may have one prefix, so try the following

    httpContext.Response.Redirect(httpContext.Request.PathBase)